### PR TITLE
feat(app): find spoken moments in search + play on open

### PR DESCRIPTION
**Pull request by aryanorastar**
## Summary
- In-app conversation search could only match **title + overview** (Typesense). Spoken phrases that live only in the transcript were invisible — the #1 retention gap vs Limitless-style “play what they said.”
- `POST /v1/conversations/search` now merges optional transcript-chunk vector hits (page 1), attaches `match_snippets` (text + start/end), and the Flutter list shows the hit then opens the transcript tab seeking/playing that moment.
- Reuses the same transcript-match helpers as MCP [#11198](https://github.com/BasedHardware/omi/pull/11198) / #6621 (shared module). Does **not** duplicate Ratnam lexical [#7421](https://github.com/BasedHardware/omi/pull/7421).

## Honest gaps
- Transcript-chunk indexing is optional (`TRANSCRIPT_CHUNK_INDEXING_ENABLED`). When off, Typesense-only hits still get snippets if the phrase appears in hydrated segments; pure spoken-word recall needs the chunk index (same as MCP).
- Audio seek needs the dense conversation artifact; if playback is unavailable we still open the transcript tab.
- Physical device smoke of seek-from-search not run in this PR (no kit).

## Test plan
- [x] `pytest tests/unit/test_mcp_transcript_search.py` (+ existing search speaker/date/pagination suites green)
- [x] OpenAPI app-client + TS/Dart/Swift clients regenerated
- [x] Automated: transcript-only phrase → timed `match_snippets` (API + helpers); Flutter `searchMomentSeekFromSnippets` opens transcript tab with start/end
- [x] Automated: locked conversations filtered from search + list/integration redaction clears `match_snippets`
- [ ] Optional maintainer smoke (needs kit + chunk index on): search spoken-only phrase → tap → audio seeks that segment

<a href="https://cubic.dev/pr/BasedHardware/omi/pull/11205?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Failure class (fixes)
Failure-Class: none

### CODE REVIEW

**Comment by aryanorastar** [comment]
## Update @2406558738

**Why “Skip” / unchecked looked scary:** CI *Skipped* jobs are path-filtered (Detect Changes → skip unused workflows). Unchecked test-plan boxes are **reviewer smokes**, not CI skips. The real blockers were Backend Unit Tests + Dart Formatting + cubic P1s.

**CI root cause:** router unit tests stub `utils.conversations` with empty `__path__`, so the new `mcp_transcript_search` submodule 404’d at import (`ModuleNotFoundError`). Stubbed it in the affected tests; also stubbed `transcript_for_llm` in KG isolation.

**Cubic:**
- P1 date one-sided → `_created_at_filter` + post-hydrate date filter
- P1 page-1 merge starvation → over-fetch → filter → slice
- P2 page>1 cost → gate transcript search to page 1
- P2 Unicode snippets → NFKC/casefold + Unicode tokens
- P2 MCP resolver unused here → stays shared for #11198
- P3 schema drift → `ConversationSearchItem` search-only

Local: previously-failed unit files + `test_mcp_transcript_search.py` green; Dart format fixed.

File: backend/routers/conversations.py [hunk 1]
```diff
@@ -1295,12 +1301,35 @@ def search_conversations_endpoint(
...
+    effective_per_page = search_results.get('per_page', 10)
+    # Spoken-word hits (optional transcript-chunk index) are merged on page 1 only so
+    # Typesense pagination stays stable. Snippets still attach for every hydrated hit.
+    transcript_ids: List[str] = []
+    if (search_request.query or '').strip():
+        transcript_ids = search_transcript_conversation_ids(
+            uid,
+            search_request.query,
+            limit=effective_per_page,
+            starts_at=start_timestamp,
```
[the remark this answers is not in the document]
Comment (aryanorastar) [reply backend/routers/conversations.py:1326]: Fixed in 2406558738: search_transcript_chunks now uses _created_at_filter (one-sided bounds), and search re-applies conversation_matches_date_range after hydration.

File: backend/routers/conversations.py [hunk 2]
```diff
@@ -1295,12 +1301,35 @@ def search_conversations_endpoint(
...
+            ends_at=end_timestamp,
+            search_transcript_chunks=vector_db.search_transcript_chunks,
+        )
+    conversation_ids = merge_typesense_page_with_transcript_hits(
```
[the remark this answers is not in the document]
Comment (aryanorastar) [reply backend/routers/conversations.py:1330]: Fixed in 2406558738: page 1 over-fetches merge candidates, hydrates/filters, then slices to per_page.

File: backend/utils/conversations/mcp_transcript_search.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    except Exception as e:  # noqa: BLE001 - transcript index is optional / best-effort
+        logger.warning(
+            "conversation search: transcript chunk search failed uid=%s: %s",
+            uid,
+            e,
+        )
+    return merge_summary_and_transcript_ids(transcript_ids, [], limit)
+
+
+def resolve_mcp_conversation_search_ids(
```
[the remark this answers is not in the document]
Comment (aryanorastar) [reply backend/utils/conversations/mcp_transcript_search.py:196]: Kept for shared helpers with #11198 (MCP wiring lands there). This PR production path is POST /v1/conversations/search.

File: backend/utils/conversations/mcp_transcript_search.py [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+from __future__ import annotations
+
+import logging
+import re
+from typing import Any, Callable, Dict, List, Optional, Sequence
+
+logger = logging.getLogger(__name__)
+
+_TOKEN_RE = re.compile(r"[a-z0-9]+", re.IGNORECASE)
```
[the remark this answers is not in the document]
Comment (aryanorastar) [reply backend/utils/conversations/mcp_transcript_search.py]: Fixed in 2406558738: Unicode-aware tokenization + NFKC/casefold.

File: backend/routers/conversations.py [hunk 5]
```diff
@@ -1295,12 +1301,35 @@ def search_conversations_endpoint(
...
     except ConversationSearchUnavailableError:
         raise HTTPException(status_code=503, detail="Search temporarily unavailable")
-    conversation_ids = [item.get('id') for item in search_results.get('items', []) if item.get('id')]
+    typesense_ids = [item.get('id') for item in search_results.get('items', []) if item.get('id')]
+    effective_page = search_results.get('current_page', 1)
+    effective_per_page = search_results.get('per_page', 10)
+    # Spoken-word hits (optional transcript-chunk index) are merged on page 1 only so
+    # Typesense pagination stays stable. Snippets still attach for every hydrated hit.
+    transcript_ids: List[str] = []
+    if (search_request.query or '').strip():
```
[the remark this answers is not in the document]
Comment (aryanorastar) [reply backend/routers/conversations.py]: Fixed in 2406558738: transcript vector search gated to effective_page == 1.

File: backend/models/conversation.py [hunk 6]
```diff
@@ -213,6 +226,9 @@ class Conversation(BaseModel):
     client_device_id: Optional[str] = None
     client_platform: Optional[str] = None
 
+    # Search-only: grep-style transcript evidence for find-and-play.
+    match_snippets: List[TranscriptMatchSnippet] = []
```
[the remark this answers is not in the document]
Comment (aryanorastar) [reply backend/models/conversation.py]: Fixed in 2406558738: moved to ConversationSearchItem (search-only); regenerated OpenAPI + clients.

**Comment by aryanorastar** [comment]
## Test plan completed (automated) @deb654b693

- [x] `test_mcp_transcript_search.py` + search speaker/date suites
- [x] OpenAPI clients (prior)
- [x] Transcript-only phrase → timed `match_snippets` (API + helpers)
- [x] Locked never leak snippets (search filter + list/integration redaction)
- [x] Flutter `searchMomentSeekFromSnippets` unit tests (tap → transcript tab + start/end)

Left unchecked on purpose: physical audio seek smoke (needs kit + chunk index on).

**Review by Git-on-my-level — DISMISSED** [review]
Thanks for the thoughtful work here — the feature direction is promising, and I like that the PR explicitly preserves locked-conversation redaction for `match_snippets`.

I can't mark this ready yet because the current head still has merge-blocking validation failures:

- `Formatting` reports `app/lib/pages/conversation_detail/page.dart` would be changed by `dart format --line-length 120`.
- `Backend unit suite` reports failures in `tests/unit/test_app_client_dart_generator.py` and `tests/unit/test_desktop_rest_inventory.py`. From the diff, the search response now intentionally exposes `ConversationSearchItem`, but the existing app-client/desktop contract tests still assert `SearchConversationsResponse.items` is `Conversation`. Please update the generator/desktop contract expectations (or the API shape, if the generated clients are not supposed to see a distinct search item model) so the OpenAPI-generated clients and first-party desktop decoder contract agree.

Because this exposes transcript-derived snippets and seek times through the app-client search API, I'd also like a maintainer/product pass before merge on the privacy/UX boundary: locked rows must continue to suppress snippets, and the app behavior should be acceptable when audio playback is unavailable or the transcript-chunk index is disabled.

Once formatting and the generated-client/desktop contract tests are green, this looks much closer.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Comment by aryanorastar** [comment]
Thanks @Git-on-my-level — merge-blocking items fixed in `33d42436f1`:

1. **Dart format** — `app/lib/pages/conversation_detail/page.dart` formatted (`dart format --line-length 120`).
2. **Contract tests** — updated expectations to `ConversationSearchItem` for `SearchConversationsResponse.items` in `test_app_client_dart_generator.py` + `test_desktop_rest_inventory.py` (intentional search-only shape with optional `match_snippets`).

Local: those two previously-failing assertions now pass.

**Privacy / UX boundary (for product pass):**
- Locked conversations are filtered out of search results; list/integration redaction clears `match_snippets` (covered by automated tests).
- Chunk index off → Typesense/title-overview hits still work; snippets only when the phrase appears in hydrated segments (honest gap in PR body).
- Audio unavailable → still opens transcript tab at the snippet seek window (no hard fail).

Physical seek-from-search smoke still not run (no kit). Ready for re-pass + privacy/UX glance when you have a moment.

**Comment by aryanorastar** [comment]
one remaining CI miss after `33d42436f1`: **Formatting** wanted black on `backend/tests/unit/test_conversation_search_date_validation.py` (line-length 120). Pushed that reformatting.

Backend unit + Dart format + ConversationSearchItem contracts remain green from the prior tip. Ready for re-pass + privacy/UX glance when CI is green.

**Review by kodjima33 — APPROVED** [review]
Spoken-moment search + play-on-open — feature, approve only.

**Comment by aryanorastar** [comment]
@Git-on-my-level **need human response** — please **dismiss the stale CHANGES_REQUESTED** on #11205.

Author-clear on your blockers:
- Dart format + `ConversationSearchItem` / desktop contract expectations were fixed on `78847a4f9f`
- @kodjima33 **APPROVED**
- Tip @`c12ce90a2f1f` = merge of current `main` + restore of SearchItem contracts (main overwrite had reset the test expectations to bare `Conversation`)
- CI was fully green before the refresh; focused contracts + `test_mcp_transcript_search` re-checked locally

Privacy/UX (for the product glance you asked for):
- Locked conversations filtered from search; redaction clears `match_snippets` (automated)
- Chunk index off → Typesense path still works; snippets only when phrase is in hydrated segments
- Audio unavailable → opens transcript tab at seek window (no hard fail)
- Honest gap: no physical seek-from-search smoke (no kit)

Please dismiss CR / re-pass when you can so the APPROVE can clear the merge gate.

**Comment by aryanorastar** [comment]
Follow-up: tip now correctly on @`b1ac42ca423e` (previous ping referenced a detached rebuild that had not pushed).

Contains: merge of current `main` + `ConversationSearchItem` contract restore if main overwrote the tests. Focused generator/desktop contracts green locally.

**Comment by Git-on-my-level** [comment]
Thanks for the follow-up fixes. I re-checked the current head and I’m leaving this as a positive signal rather than a formal automation approval.

What I verified:
- `backend/routers/conversations.py` now merges transcript-chunk hits only on page 1, preserves result ordering after hydration, re-applies locked/speaker/date filters before snippets are attached, and trims back to the effective page size.
- `backend/utils/conversations/mcp_transcript_search.py` is fail-open around optional transcript chunk search and builds bounded snippet evidence with segment timing for seek-to-moment behavior.
- `backend/utils/conversations/render.py` clears `match_snippets` for locked list/integration rows, which addresses the privacy boundary from the earlier review.
- `app/lib/backend/http/api/conversations.dart` and `app/lib/backend/schema/conversation.dart` now keep `match_snippets` through the generated search-item shape without widening the base generated conversation model.
- `app/lib/pages/conversation_detail/page.dart` and `app/lib/pages/conversations/widgets/conversation_list_item.dart` wire the first timed snippet into the transcript tab and tolerate unavailable playback by still opening the transcript.
- `backend/tests/unit/test_app_client_dart_generator.py`, `backend/tests/unit/test_desktop_rest_inventory.py`, and the generated OpenAPI/TS/Dart client files now reflect `ConversationSearchItem`, so the earlier generated-client/desktop contract blocker is resolved.

The current GitHub checks are green, and I also smoke-checked the new pure transcript-search helper behavior locally. I did not find a new blocking issue in the sampled code paths. Because this touches transcript-derived user data and app search UX, I’m relying on the existing human maintainer sign-off for the product/privacy boundary rather than issuing a formal automation approval.

---
*by AI on behalf of David — positive signal only; transcript-search privacy/UX has human maintainer sign-off on this head.*

**Comment by aryanorastar** [comment]
@kodjima33 @beastoin need human response — merge call for #11205 (spoken-moment search + play-on-open).

Head `@b1ac42ca42` · **APPROVED** · **MERGEABLE CLEAN** · CI green · David dismissed formatting/contract CR · `positive-signal` (no remaining production-code blocker; left privacy/UX for human maintainer).

Author-clear privacy/UX for that glance:
- Locked conversations filtered from search; list/integration redaction clears `match_snippets`
- Chunk index off → Typesense path still works; snippets only when phrase is in hydrated segments
- Audio unavailable → opens transcript tab at seek window (no hard fail)
- Honest gap: no physical seek-from-search smoke (no kit)

Please merge (or state hold) when ready — retention wow lane.

**Comment by aryanorastar** [comment]
@kodjima33 @beastoin need human response — merge call for #11205 (spoken-moment search + play-on-open).

Author-clear on `b1ac42ca42`:

- **APPROVED** (kodjima33) + **MERGEABLE CLEAN** + all checks green (24 success)
- David dismissed formatting/contract CR; `positive-signal`; no remaining production-code blocker
- Privacy/UX for the glance: locked rows filtered + `match_snippets` redacted; chunk index optional; audio unavailable → transcript tab at seek window
- Honest gap: no physical seek-from-search smoke (no kit)

Please merge when ready (or state an explicit hold) — retention wow lane.

**Comment by Git-on-my-level** [comment]
Follow-up review on the current head/check state:

The implementation still looks directionally sound to me, and I’m not adding a blocking review over the existing maintainer approval. I did re-check the main code paths:

- `backend/routers/conversations.py` merges transcript-chunk hits only into page 1, preserves hydrated result order, then re-applies locked/speaker/date filtering before `match_snippets` are attached.
- `backend/utils/conversations/mcp_transcript_search.py` keeps transcript index lookup best-effort/fail-open and bounds snippet construction to a small set of timed segments.
- `backend/utils/conversations/render.py` clears `match_snippets` for locked list/integration redaction, so the new search evidence does not bypass the locked-conversation privacy boundary.
- `app/lib/backend/http/api/conversations.dart` plus `app/lib/backend/schema/conversation.dart` preserve the generated `ConversationSearchItem.match_snippets` through to `ServerConversation` without changing the base generated conversation shape.
- `app/lib/pages/conversations/widgets/conversation_list_item.dart` and `app/lib/pages/conversation_detail/page.dart` wire the first timed snippet into transcript-tab open/seek behavior and tolerate unavailable playback.
- The generated OpenAPI/TS/Dart clients and focused tests (`backend/tests/unit/test_mcp_transcript_search.py`, `backend/tests/unit/test_conversation_search_date_validation.py`, `app/test/unit/search_moment_seek_test.dart`) cover the new contract and privacy edge cases.

Current merge blocker I see is CI, not a new product-code blocker: `PR Metadata Preflight` is now failing the product-file line-count ratchet for `backend/routers/conversations.py` against the current `origin/main` baseline, and GitHub also shows the `Formatting` check failing. Please fix/rerun those before merge; the transcript-search privacy/UX boundary still deserves human maintainer sign-off because this feature exposes transcript-derived search evidence and seek behavior.

---
*by AI on behalf of David — follow-up comment only; not a formal approval or merge decision.*

**Comment by aryanorastar** [comment]
Superseded by clean replacement PR #11400. The implementation was rebased onto current `main` as a single-parent commit to remove the stale GitHub conflict state; the verified final tree and focused tests are unchanged.
